File: leetcode/3. Longest Substring Without Repeating Characters/Concise/concise.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        d = {}
        cur, best = 0, 0
        for i in range(len(s)):
            if s[i] in d and d[s[i]] >= i-cur:
                best = max(best, cur)
                cur = i-d[s[i]]
            else:   
                cur += 1
            d[s[i]] = i
        return max(best, cur)

File: leetcode/3. Longest Substring Without Repeating Characters/Concise/concise.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map <char, int> m;
        int cur = 0;
        int best = 0;
        for(int i=0; i<s.length(); i++){
            if(m.count(s[i])>0 && m[s[i]]>=i-cur){
                if(cur>best){
                    best = cur;
                }
                cur = i - m[s[i]];
            } else {
                cur += 1;
            }
            m[s[i]] = i;
        }
        if(cur>best){
            best = cur;
        }
        return best;
    }
};